include/drivers/i2c.h

```diff
@@ -337,6 +337,166 @@ struct i2c_slave_config {
 	const struct i2c_slave_callbacks *callbacks;
 };
 
+#if defined(CONFIG_I2C_STATS) || defined(__DOXYGEN__)
+
+#include <stats/stats.h>
+
+STATS_SECT_START(i2c)
+STATS_SECT_ENTRY32(bytes_read)
+STATS_SECT_ENTRY32(bytes_written)
+STATS_SECT_ENTRY32(message_count)
+STATS_SECT_ENTRY32(transfer_call_count)
+STATS_SECT_END;
+
+STATS_NAME_START(i2c)
+STATS_NAME(i2c, bytes_read)
+STATS_NAME(i2c, bytes_written)
+STATS_NAME(i2c, message_count)
+STATS_NAME(i2c, transfer_call_count)
+STATS_NAME_END(i2c);
+
+/**
+ * @brief I2C specific device state which allows for i2c device class specific additions
+ */
+struct i2c_device_state {
+	struct device_state devstate;
+	struct stats_i2c stats;
+};
+
+/**
+ * @brief Updates the i2c stats for i2c transfers
+ */
+static inline void i2c_xfer_stats(const struct device *dev, struct i2c_msg *msgs,
+				  uint8_t num_msgs)
+{
+	struct i2c_device_state *state =
+		CONTAINER_OF(dev->state, struct i2c_device_state, devstate);
+	uint32_t bytes_read = 0U;
+	uint32_t bytes_written = 0U;
+
+	STATS_INC(state->stats, transfer_call_count);
+	STATS_INCN(state->stats, message_count, num_msgs);
+	for (uint8_t i = 0U; i < num_msgs; i++) {
+		if (msgs[i].flags & I2C_MSG_READ) {
+			bytes_read += msgs[i].len;
+		}
+		if (msgs[i].flags & I2C_MSG_WRITE) {
+			bytes_written += msgs[i].len;
+		}
+	}
+	STATS_INCN(state->stats, bytes_read, bytes_read);
+	STATS_INCN(state->stats, bytes_written, bytes_written);
+}
+
+/**
+ * @brief Define a statically allocated and section assigned i2c device state
+ */
+#define I2C_DEVICE_STATE_DEFINE(node_id, dev_name)	\
+	static struct i2c_device_state Z_DEVICE_STATE_NAME(dev_name)	\
+	__attribute__((__section__(".z_devstate")));
+
+/**
+ * @brief Define an i2c device init wrapper function
+ *
+ * This does device instance specific initialization of common data (such as stats)
+ * and calls the given init_fn
+ */
+#define I2C_INIT_FN(dev_name, init_fn)					\
+	static inline int UTIL_CAT(dev_name, _init)(const struct device *dev) \
+	{								\
+		struct i2c_device_state *state =			\
+			CONTAINER_OF(dev->state, struct i2c_device_state, devstate); \
+		stats_init(&state->stats.s_hdr, STATS_SIZE_32, 4,	\
+			   STATS_NAME_INIT_PARMS(i2c));			\
+		stats_register(dev->name, &(state->stats.s_hdr));	\
+		return init_fn(dev);					\
+	}
+
+#define _I2C_DEVICE_DT_DEFINE(node_id, init_fn, pm_device,		\
```

Comment:
ditto

Reply:
note that this auxiliary could be deleted:

```c
#ifdef CONFIG_I2C_STATS
#define I2C_DEVICE_DT_DEFINE(...) /* the stats version */
#else
#define I2C_DEVICE_DT_DEFINE(...) /* the non-stats version */
#endif
```

and document just one (mentioning what happens when `@kconfig{CONFIG_I2C_STATS}` is enabled.